### PR TITLE
feat: add mtls feature to rest transport

### LINKED ISSUE #383: support mTLS for http
when http is supported by microgenerator.

### PR DESCRIPTION
**Pull request by arithmetic1728**
Currently only grpc transport support mtls. This PR adds mtls support to http transport and fixes #383. The logic is the same. The only difference is we use different apis for different transports.

The following table shows the apis used for certificates.
| transport | get user provided cert | check default cert existence | get default cert | add cert to channel |
| :--- | :--- | :--- | :--- | :--- |
| grpc | client_options.client_cert_source | SslCredentials().is_mtls | SslCredentials().ssl_credentials | use grpc.ssl_channel_credentials created from client_cert_source |
| http | client_options.client_cert_source |  mtls.has_default_client_cert_source() | mtls.default_client_cert_source() | just call AuthorizedSession.configure_mtls_channel(client_cert_source) method |

The following code (extracted from the PR) demonstrates the api differences.

http (new code from this PR)
```
client_cert_source = None
is_mtls = False
if use_client_cert:
    if client_options.client_cert_source:
        is_mtls = True
        client_cert_source = client_options.client_cert_source                
    else:
        is_mtls = mtls.has_default_client_cert_source()
        client_cert_source = mtls.default_client_cert_source() if is_mtls else None
```

grpc (existing code)
```
ssl_credentials = None
is_mtls = False
if use_client_cert:
    if client_options.client_cert_source:
        import grpc  # type: ignore
        cert, key = client_options.client_cert_source()
        ssl_credentials = grpc.ssl_channel_credentials(
            certificate_chain=cert, private_key=key
        )
        is_mtls = True
    else:
        creds = SslCredentials()
        is_mtls = creds.is_mtls
        ssl_credentials = creds.ssl_credentials if is_mtls else None
```

### CODE REVIEW

**Review by busunkim96 — APPROVED** [review]

**Comment by yon-mg** [comment]
Thanks for doing this!

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 1]
```diff
@@ -292,7 +304,11 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
                 credentials_file=client_options.credentials_file,
                 host=api_endpoint,
                 scopes=client_options.scopes,
+                {%- if 'rest' in opts.transport %}
+                client_cert_source=client_cert_source,
+                {% else %}
                 ssl_channel_credentials=ssl_credentials,
+                {% endif %}
```
[the remark this answers is not in the document]
Comment (arithmetic1728) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:311]: I see. Then it seems I need to create both `ssl_channel_credentials` for grpc, and `client_cert_source` for http, and pass both to the transport constructor, and let the constructor decide which one to use. I will make the changes.

[the remark this answers is not in the document]
Comment (arithmetic1728) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2]: I changed the code. Now the logic is:
1. From client side, we will use `client_cert_source` for mtls regardless of the transport. The `client_cert_source` is either the user provided one from `client_options`, or the default one from `mtls.default_client_cert_source()`. Client passes `client_cert_source` to transport via a new `client_cert_source_for_mtls` argument. 
2. Rest transport will use `client_cert_source_for_mtls` to configure mtls.
3. Grpc transport will use `ssl_channel_credentials` if provided, and if not then use `client_cert_source_for_mtls`. So this won't bring any breaking changes.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 2]
```diff
@@ -239,21 +239,15 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
         # Create SSL credentials for mutual TLS if needed.
         use_client_cert = bool(util.strtobool(os.getenv("GOOGLE_API_USE_CLIENT_CERTIFICATE", "false")))
 
-        ssl_credentials = None
+        client_cert_source = None
```
Comment (software-dov) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:242]: Minor naming nit: maybe `client_cert_source_func` to make it clear that this is a callable?

**Review by software-dov — APPROVED** [review]
LGTM, just a minor naming nit.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 (hunk 2 above)
Comment (arithmetic1728) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2]: done

**Pull request merged**